language-design.md

```diff
@@ -28,7 +75,69 @@ you know a poorly designed language? What are the symptoms?
+allow for growth, but allowing the language to be modified by people other than 
+the creators allows for a bigger community which makes the language more 
+supported. Steele talks about this extensively in his article, "Growing a 
+Language" [Steele, 1998]. A language that does not get a following or have a 
+community will soon die. Though one cannot definitely say a dead language or 
+one that is unused / used by very people means it is a badly designed language, 
+we think there is definitely a correlation. Steele touches on this by talking 
+about Java and how he and others that work on Java do not define the way it is 
+going, but rather a "way of doing things". 
+
+Another symptom of a well-defined language is that it has well-defined 
+semantics / language building blocks. This contributes to the elements above, 
+specifically, the easiness to read and write. More importantly, well-defined 
+semantics and building blocks makes the language predictable. As Jean Yang says, 
+a good programming language should have well-defined semantics that "make it 
+easy to reason about what your program will do" [Yang, 2012].
+
+For languages use in industry, there are a few more elements that contribute to 
+a well-designed language. Nystrom, in his article about what makes a well-
+designed languages, talks about how a well-designed languages should be 
+scalable for large systems and teams over time [Nystrom, 2011]. These are some 
+things we have not talked about in class, but are very important when choosing 
+languages in industry. For example, Go is seeing a rapid rise in popularity in 
+industry because of the easiness of doing concurrency in Go. 
+
+For DSLs, there are even more considerations. "Well-known GPL design criteria 
+such as readability, simplicity, orthogonality,the design principles listed by 
+Brooks[1996], and Tennent’s design principles [1977] retain some validity for 
+DSLs. However, the DSL designer has to keep in mind both the special character 
+of DSLs as well as the fact that users need not be programmers" [Mernik et al., 2005]. 
+Mernik et al. go on to talk more about the specifics of good DSL design, but this 
+quote is important as it talks about how when designing a DSL, one should be 
+aware of the language's audience, and the nature of the domain the language is 
+for. 
+
+There are also many things that make a language badly-designed. Some obvious 
+symptoms are bad naming conventions and contradicting popular conventions.
+If a language's key words contradict those used by popular language, for 
+example, if "break" in some language means exposing elements from an array 
+instead of interrupting a loop procedure, then it can be seen as bad language 
+design. 
+
+If a langauge has ambiguities and uncertainties, then it is also badly 
+designed. Though JavaScript is very widely used, it is some ambiguities that 
+make it a badly-designed language. For example, the "this" keyword is used 
```

Comment:
So true. This is one of the biggest reasons why Javascript annoys me, pun intended. Debugging is also hard in general (at least more so than other programming languages).